homeassistant/components/camera/android_ip_webcam.py

```diff
@@ -0,0 +1,128 @@
+
+import aiohttp
+import async_timeout
+import requests
+from requests.auth import HTTPBasicAuth
+
+from homeassistant.components.camera import (PLATFORM_SCHEMA, Camera)
+from homeassistant.helpers.aiohttp_client import (
+    async_get_clientsession, async_aiohttp_proxy_stream)
+from homeassistant.components import android_ip_webcam
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({})
+
+DEPENDENCIES = ['android_ip_webcam']
+
+
+@asyncio.coroutine
+# pylint: disable=unused-argument
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Setup an IP Webcam Camera."""
+    if discovery_info is None:
+        return
+    devices = hass.data[android_ip_webcam.DATA_IP_WEBCAM]
+    cameras = [IPWebcamCamera(hass, device)
+               for key, device in devices.items()]
+    yield from async_add_devices(cameras, True)
+
+
+def extract_image_from_mjpeg(stream):
+    """Take in a MJPEG stream object, return the jpg from it."""
+    data = b''
+    for chunk in stream:
+        data += chunk
+        jpg_start = data.find(b'\xff\xd8')
+        jpg_end = data.find(b'\xff\xd9')
+        if jpg_start != -1 and jpg_end != -1:
+            jpg = data[jpg_start:jpg_end + 2]
+            return jpg
+
+
+class IPWebcamCamera(Camera):
+    """An implementation of an IP camera that is reachable over a URL."""
+
+    def __init__(self, hass, device):
+        """Initialize a IP Webcam camera."""
+        super(IPWebcamCamera, self).__init__()
+        self._device = device
+        self._name = self._device.name
+        self._username = self._device.username
+        self._password = self._device.password
+        self._mjpeg_url = '{}/{}'.format(self._device.base_url, 'video')
+        self._still_image_url = '{}/{}'.format(self._device.base_url,
+                                               'photo.jpg')
+
+        self._auth = None
+        if self._username and self._password:
+            self._auth = aiohttp.BasicAuth(self._username,
+                                           password=self._password)
+
+    @asyncio.coroutine
+    def async_camera_image(self):
+        """Return a still image response from the camera."""
+        websession = async_get_clientsession(self.hass)
+        response = None
+        try:
+            with async_timeout.timeout(10, loop=self.hass.loop):
+                response = yield from websession.get(
+                    self._still_image_url, auth=self._auth)
+
+                image = yield from response.read()
+                return image
+
+        except asyncio.TimeoutError:
+            _LOGGER.error('Timeout getting camera image')
+
+        except (aiohttp.errors.ClientError,
+                aiohttp.errors.ClientDisconnectedError) as err:
+            _LOGGER.error('Error getting new camera image: %s', err)
+
+        finally:
+            if response is not None:
+                yield from response.release()
+
+    def camera_image(self):
```

Comment:
Oh it's necessary in `mjpeg.py` in case users have Digest authentication or don't have a still image URL set.